File: src/main/java/gargoyle/netsquares/model/player/HumanPlayer.java
package gargoyle.netsquares.model.player;

import gargoyle.netsquares.model.*;
import gargoyle.netsquares.ui.listener.NSKeyListener;
import gargoyle.netsquares.ui.listener.NSMouseListener;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

public class HumanPlayer extends Player implements NSKeyListener, NSMouseListener {
    private final Collection<Action> persistentActions = new LinkedHashSet<>();
    private final Queue<Action> stackedActions = new ArrayDeque<>();
    private KeyMap keyMap = new KeyMap();

    public HumanPlayer(String name) {
        super(name);
    }

    public KeyMap getKeyMap() {
        return keyMap;
    }

    public void setKeyMap(KeyMap keyMap) {
        this.keyMap = keyMap;
    }

    @SuppressWarnings("LawOfDemeter")
    @Override
    protected Move move(@NotNull Board board, @NotNull Directions directions) {
        if (stackedActions.isEmpty()) {
            Iterator<Action> iterator = persistentActions.iterator();
            if (iterator.hasNext()) {
                Action action = iterator.next();
                Move move = action.getMove();
                if (move.isAllowed(directions)) {
                    if (move == Move.OPEN) {
                        persistentActions.clear();
                    }
                    return move;
                }
            }
            return null;
        } else {
            return stackedActions.poll().getMove();
        }
    }

    @Override
    public void onClick(Action[] actions) {
        stackedActions.addAll(Arrays.asList(actions));
    }

    @Override
    public void onKeyPress(@Nullable Action action) {
        if (action != null) {
            persistentActions.add(action);
        }
    }

    @Override
    public void onKeyRelease(@Nullable Action action) {
        if (action != null) {
            persistentActions.remove(action);
        }
    }
}

File: src/main/kotlin/gargoyle/netsquares/model/player/HumanPlayer.kt
package gargoyle.netsquares.model.player

import gargoyle.netsquares.model.Action
import gargoyle.netsquares.model.Board
import gargoyle.netsquares.model.Directions
import gargoyle.netsquares.model.KeyMap
import gargoyle.netsquares.model.Move
import gargoyle.netsquares.model.Player
import gargoyle.netsquares.ui.listener.NSKeyListener
import gargoyle.netsquares.ui.listener.NSMouseListener
import java.util.ArrayDeque
import java.util.Queue

class HumanPlayer(name: String) : Player(name), NSKeyListener, NSMouseListener {
    private val persistentActions: MutableCollection<Action> = LinkedHashSet()
    private val stackedActions: Queue<Action> = ArrayDeque()
    var keyMap = KeyMap()
    override fun move(board: Board, directions: Directions?): Move? {
        return if (stackedActions.isEmpty()) {
            val iterator: Iterator<Action> = persistentActions.iterator()
            if (iterator.hasNext()) {
                val action = iterator.next()
                val move = action.move
                if (move.isAllowed(directions)) {
                    if (move == Move.OPEN) {
                        persistentActions.clear()
                    }
                    return move
                }
            }
            null
        } else {
            stackedActions.poll().move
        }
    }

    override fun onClick(actions: Array<Action?>) {
        stackedActions.addAll(listOf(*actions))
    }

    override fun onKeyPress(action: Action?) {
        if (action != null) {
            persistentActions.add(action)
        }
    }

    override fun onKeyRelease(action: Action?) {
        if (action != null) {
            persistentActions.remove(action)
        }
    }
}
